### PR TITLE
TASK: Make E2E tests work for Neos 9.0

### PR DESCRIPTION
**Pull request by grebaldi**
refs: #3546, #3406
~includes: #3571~
~includes: https://github.com/neos/neos-ui/pull/3618~

This is the third and final in a series of PRs leading up to Neos 9.0. I want to make sure that our E2E tests can be executed locally on every maintained branch.

**How to test this**

You should of course be able to run the E2E tests locally via
```sh
make test-e2e-docker browser=chromium # or browser=chrome
```

The E2E tests use `cweagans/composer-patches` to include patches from the development collection. Documentation can be found here: https://github.com/cweagans/composer-patches/blob/1.x/README.md
We discussed to keep this dependency to make developing across repos easier ;)

**Remaining TODOs**

- [x] Make anything run at all
- [x] Fixture: 1Dimension
  - [x] Create event export from site export
  - [x] Fix all occurring E2E test failures
    - [x] 1Dimension/createNewNodeInContainer.e2e.js Line 33: "AssertionError: expected false to be truthy"
      - Root cause is here: https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Resources/Private/Fusion/Prototypes/ContentCollection.fusion#L30
      - I suppose that both `Neos.Node.inBackend` and `node.context.currentRenderingMode.edit` can no longer be relied upon that easily
      - Known as: https://github.com/neos/rector/issues/23
      - [x] Most likely fixed by: https://github.com/neos/neos-development-collection/pull/4067
    - [x] 1Dimension/discarding.e2e.js Line 41: "AssertionError: After discarding we are back to the main page: expected 'user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__d7315383-adb8-45a5-8124-ffabeee
bc51a' to deeply equal '/sites/neos-test-site@user-admin;language=en_US'"
       - [x] Find proof that `/sites/neos-test-site@user-admin;language=en_US` and `user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__d7315383-adb8-45a5-8124-ffabeeebc51a` represent the same node
         - They don't. The expected node address should rather be `user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__f676459d-ca77-44bc-aeea-44114814c279`. `d7315383-adb8-45a5-8124-ffabeeebc51a` presumably refers to the node that has just been discarded in this test. Redirect to the home page ought to happen, but fails.
       - There is also an error: `TODO: SITE NOT FOUND; should not happen (for address NodeAddress[contentStream=9ce2581c-6e0a-467a-ad33-37c7133ce3a2, dimensionSpacePoint={"language":"en_US"}, nodeAggregateId=6b0b30db-933b-4c3d-b47b-0a7beed6c315, workspaceName=user-admin]`
       - This happens because the guest frame is trying to render the discarded node
       - Nicer error message would need to be solved in https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Classes/Controller/Frontend/NodeController.php#L146 and https://github.com/neos/neos-development-collection/blob/9.0/Neos.Neos/Classes/Controller/Frontend/NodeController.php#L222
       - Known as: https://github.com/neos/neos-development-collection/issues/4024
       - May be fixed by: https://github.com/neos/neos-development-collection/pull/4393
     - [x]  1Dimension/discarding.e2e.js Line 61: "AssertionError: Discarded node gone from the tree: expected true to be falsy"
       - Same "SITE NOT FOUND" error
     - [x] 1Dimension/switchingDimensions.e2e.js Line 24: "AssertionError: Untranslated node gone from the tree: expected true to be falsy"
       - [x] Will be fixed with: https://github.com/neos/neos-development-collection/pull/4390
     - [x] 1Dimension/switchingDimensions.e2e.js Line 23: "getReact cannot return DOM elements. Use Selector functions for this purpose."
       - After switching dimensions, the focused page tree "contextPath" ends up being a DOM node:  image
       - Regression through https://github.com/neos/neos-ui/pull/3173 - can be easily fixed
     - [x] 1Dimension/treeMultiselect.e2e.js Line 29: "AssertionError: Node B's context path changed: expected 'user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__5b0d6ac0-40ab-47e8-b79e-39de6c0700df' to deeply
equal '/sites/neos-test-site/node-knm2pltb5454z/node-18qsaeidy6765/node-e8tw6sparbtp3@user-admin;language=en_US'"
        - [x] Find proof that `/sites/neos-test-site/node-knm2pltb5454z/node-18qsaeidy6765/node-e8tw6sparbtp3@user-admin;language=en_US` and `user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__5b0d6ac0-40ab-47e8-b79e-39de6c0700df` represent the same node
          - https://github.com/neos/neos-ui/blob/bc6569fd04180e650d98f8f1c7d3d214593e951d/Tests/IntegrationTests/Fixtures/1Dimension/SitePackage/Resources/Private/Content/Sites.xml#L472
      - [x] 1Dimension/treeMultiselect.e2e.js Line 46: "AssertionError: Node B's context path changed: expected 'user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__5b0d6ac0-40ab-47e8-b79e-39de6c0700df' to deeply
equal '/sites/neos-test-site/node-knm2pltb5454z/node-18qsaeidy6765/node-e8tw6sparbtp3@user-admin;language=en_US'"
        - [x] Find proof that `/sites/neos-test-site/node-knm2pltb5454z/node-18qsaeidy6765/node-e8tw6sparbtp3@user-admin;language=en_US` and `user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__5b0d6ac0-40ab-47e8-b79e-39de6c0700df` represent the same node
          - https://github.com/neos/neos-ui/blob/bc6569fd04180e650d98f8f1c7d3d214593e951d/Tests/IntegrationTests/Fixtures/1Dimension/SitePackage/Resources/Private/Content/Sites.xml#L472
      - [x] 1Dimension/treeMultiselect.e2e.js Line 63: "AssertionError: Node C's context path changed: expected 'user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__84eb0340-ba34-4fdb-98b1-da503f967121' to deeply
equal '/sites/neos-test-site/node-knm2pltb5454z/node-18qsaeidy6765/node-oml0cxaompt29@user-admin;language=en_US'"
        - [x] Find proof that `/sites/neos-test-site/node-knm2pltb5454z/node-18qsaeidy6765/node-oml0cxaompt29@user-admin;language=en_US` and `user-admin__eyJsYW5ndWFnZSI6ImVuX1VTIn0=__84eb0340-ba34-4fdb-98b1-da503f967121` represent the same node
          - https://github.com/neos/neos-ui/blob/bc6569fd04180e650d98f8f1c7d3d214593e951d/Tests/IntegrationTests/Fixtures/1Dimension/SitePackage/Resources/Private/Content/Sites.xml#L499
- [x] Fixture: 2Dimension
  - [x] Create event export from site export
  - [x] Fix all occurring E2E test failures
    - [x] 2Dimension/switchingDimensions.e2e.js Line 23: "A JavaScript error occurred on '...'."
      - Also, an error flash-message appears: image
      - Solution lives here: https://docs.neos.io/guide/manual/routing/configuring-dimension-resolving
      - After adding config for dimension resolution, I get a weird thing: image
      - Cause: `Router route(): No route matched the route path "neos/service/content-dimensions/country.json"`
      - Regression through https://github.com/neos/neos-development-collection/pull/3848
      - [x] Possibly fixed by: https://github.com/neos/neos-development-collection/pull/4395
- [x] Merge both fixtures as described in neos/neos-development-collection#3406
  - [x] Create unified test distribution with both, the 1Dimension and the 2Dimension case, as site packages with separate content repositories
    - This requires https://github.com/neos/neos-development-collection/pull/4400 to work
  - [x] Add test for switching between sites
    - <s>We're running into https://github.com/DevExpress/testcafe/issues/7548 on this one</s>
      - Nope, but *are* running into https://github.com/neos/neos-development-collection/issues/4401, but only if the switch sites test is executed in isolation (because otherwise, `switchDimensions` runs first, which creates the missing user workspace)
    - <s>Update Testcafé to v3 and use proxyless mode to execute tests</s>
      - I abandoned this, even though I actually did the update and it was quite easy. But since the Testcafé version wasn't the cause for the issue that I've seen for the "switching between sites" test, and the update breaks other tests, I've decided to postpone Testcafé v3 for now. However, this can easily be done in a separate PR :)
  - [x] Use better strategy for local docker setup, now that no more package switching needs to be done
- [x] Make E2E tests work on Circle CI
- [x] Restore PHP Unit Tests on Circle CI
- [x] ~Buckle up for collisions with https://github.com/neos/neos-ui/pull/3503~ -> merge this first
- [x] Remove dangling patches after PRs have been merged
  - [x] https://github.com/neos/neos-development-collection/pull/4393
  - [x] https://github.com/neos/neos-development-collection/pull/4395
  - [x] https://github.com/neos/neos-development-collection/pull/4067
    - [x] revert ce395ca9e3614ec32f7af7a82d2759a74e68f62d  when https://github.com/neos/neos-ui/pull/3411 is merged
  - [x] https://github.com/neos/neos-development-collection/pull/4400
  - [ ] revert `Neos.Flow.object.includeClasses` hack!

### CODE REVIEW

File: Tests/IntegrationTests/utils.js
Comment (mhsdesign) [inline Tests/IntegrationTests/utils.js]: sehr geil

**Comment by mhsdesign** [comment]
image

you are a hero ^^

**Review by skurfuerst — APPROVED** [review]
just looked over this - GREAT WORK <3 <3 I'll check that we get the dependent PRs merged soon, so that we can get this one in.

**Comment by mhsdesign** [comment]
As there might be a collision as you pointed out, i will take care of reviewing and merging: https://github.com/neos/neos-ui/pull/3503

Also it seems like https://github.com/neos/neos-development-collection/pull/4067 is blocking this pr from moving forward so we should definitely focus there as well ;) But we discussed in the last escr meeting a plan for the preview modes its going forward - slowly ^^

File: packages/neos-ui-backend-connector/src/Endpoints/index.ts
Comment (mhsdesign) [inline packages/neos-ui-backend-connector/src/Endpoints/index.ts:487]: why did this work previously ??? 😂

File: packages/neos-ui-sagas/src/UI/ContentTree/index.js
Comment (mhsdesign) [inline packages/neos-ui-sagas/src/UI/ContentTree/index.js:145]: this fix could be backported as well no?

(would also take care of it just vote: 👍  👎 )

File: packages/neos-ui-backend-connector/src/Endpoints/index.ts
Comment (grebaldi) [reply packages/neos-ui-backend-connector/src/Endpoints/index.ts:487]: I suppose `.node-context-path` was never empty before, but occurred to be empty during the tests, because of all the backend changes.

File: packages/neos-ui-sagas/src/UI/ContentTree/index.js
Comment (grebaldi) [reply packages/neos-ui-sagas/src/UI/ContentTree/index.js:145]: My plan was to leave the plow-js calls as they are for now and take care of replacing them in https://github.com/neos/neos-ui/pull/3306, right after the E2E tests have been merged.

**Comment by mhsdesign** [comment]
I did revert your hotfix for edit preview modes ce395ca9e3614ec32f7af7a82d2759a74e68f62d

and merged this final implementation: https://github.com/neos/neos-ui/pull/3618 into your branch

lets see what the tests say :D

**Comment by mhsdesign** [comment]
I added a hack so the tests should now work in ci :D - will take care of a proper fix.

locally `make test-e2e-docker` doesnt seem to work as i think i have problems with `http://onedimension.localhost:8081/neos` and `http://twodimensions.localhost:8081/neos` 

-> it seems one needs to modify the `/etc/hosts` ?

```
          echo 127.0.0.1 onedimension.localhost | sudo tee -a /etc/hosts
          echo 127.0.0.1 twodimensions.localhost | sudo tee -a /etc/hosts
```

so we need a tutorial for this for mac as well ;)

**Review by mhsdesign — APPROVED** [review]
I just ran the tests `make test-e2e-docker` locally on my mac and they work flawlessly!

As soon as the ci is green we can merge this :D

**Pull request merged**
